src/data/qpf/univariate/pfunctor/M.lean

```diff
@@ -0,0 +1,794 @@
+      induction h' : (f' i) using pfunctor.M.cases_on' with a₁ f₁,
+      simp only [h,h',isubtree_cons] at ps_ih ⊢,
+      rw [h,h'] at h₁,
+      have : a₀ = a₁ := bisim.head h₁, subst a₁,
+      apply (ps_ih _ _ _ h₁),
+      rw [← h,← h'], apply or_of_or_of_imp_of_imp hh is_path_cons' is_path_cons'
+  end
+
+  theorem eq_of_bisim [nonempty (M F)] (bisim : is_bisimulation R) : ∀ s₁ s₂, s₁ ~ s₂ → s₁ = s₂ :=
+  begin
+    inhabit (M F),
+    introv Hr, apply ext,
+    introv,
+    by_cases h : is_path ps s₁ ∨ is_path ps s₂,
+    { have H := nth_of_bisim R bisim _ _ ps Hr h,
+      exact H.left },
+    { rw not_or_distrib at h, cases h with h₀ h₁,
+      simp only [iselect_eq_default,*,not_false_iff] }
+  end
+end bisim
+
+section coinduction
+
+variables F
+
+/-- generic bisimulation relation for M-types -/
+coinductive R : Π (s₁ s₂ : M F), Prop
+| intro {a} (s₁ s₂ : F.B a → M F) :
+   (∀ i, R (s₁ i) (s₂ i)) →
+   R (M.mk ⟨_,s₁⟩) (M.mk ⟨_,s₂⟩)
+
+attribute [nolint doc_blame] R.functional R
+
+open ulift
+
+@[nolint def_lemma]
+lemma R_is_bisimulation : is_bisimulation (R F) :=
+begin
+  constructor; introv hr,
+  { suffices : (λ a b, head a = head b) (M.mk ⟨a, f⟩) (M.mk ⟨a', f'⟩),
+    { simp only [head_mk] at this, exact this },
+    refine R.cases_on _ hr _,
+    intros, simp only [head_mk] },
+  { suffices : (λ a b, ∀ i j, i == j → R F (children a i) (children b j)) (M.mk ⟨a, f⟩) (M.mk ⟨a, f'⟩),
+    { specialize this (cast (by rw head_mk) i) (cast (by rw head_mk) i) heq.rfl,
+      simp only [children_mk] at this, exact this, },
+    refine R.cases_on _ hr _,
+    introv h₂ h₃,
+    let k := cast (by rw head_mk) i_1,
+    have h₀ : (children (M.mk ⟨a_1, s₁⟩) i_1) = s₁ k := children_mk _ _,
+    have h₁ : (children (M.mk ⟨a_1, s₂⟩) j) = s₂ k,
+    { rw children_mk, congr, symmetry, apply eq_of_heq h₃ },
+    rw [h₀,h₁], apply h₂ },
+end
+
+variables {F}
+
+lemma coinduction {s₁ s₂ : M F}
+  (hh : R _ s₁ s₂) :
+  s₁ = s₂ :=
+begin
+  haveI := inhabited.mk s₁,
+  exact eq_of_bisim
+    (R F) (R_is_bisimulation F) _ _
+    hh
+end
+
+lemma coinduction' {s₁ s₂ : M F}
+  (hh : R _ s₁ s₂) :
+  s₁ = s₂ :=
+begin
+  have hh' := hh, revert hh',
+  apply R.cases_on F hh, clear hh s₁ s₂,
+  introv h₀ h₁,
+  rw coinduction h₁
+end
+
+end coinduction
+
+universes u' v'
+
+/-- corecursor for `M F` with swapped arguments -/
+def corec_on {X : Type*} (x₀ : X) (f : X → F.obj X) : M F :=
+M.corec f x₀
+
+end M
+
+end pfunctor
+
+namespace tactic.interactive
+open tactic (hiding coinduction) lean.parser interactive interactive.types
+
+/-- tactic for proving equality on `M F` through bisimulation -/
+meta def bisim (ids : parse $ types.with_ident_list) (g : parse $ optional (tk "generalizing" *> many ident)) : tactic unit :=
+do applyc ``pfunctor.M.coinduction,
+   coinduction ``pfunctor.M.R.corec_on ids g
+
+end tactic.interactive
+
+namespace pfunctor
+
+open M
+
+variables {P : pfunctor.{u}} {α : Type u}
+
+/-- re-export M.dest -/
```

Comment:
I think that `M.dest` is a perfectly fine name and doesn't require re-exporting.  If you still want to add an extra definition (which is probably suboptimal because of the extra unfolding), please copy the doc string.